File: compiler/src/typecheck/constraints/type_constraint.rs
use crate::{
    database::NodeRef,
    typecheck::{
        Constraint, ConstraintCtx, ConstraintInfo, ConstraintResult, ConstructedType,
        GroupConstraint, InstantiateContext, Type,
    },
};

#[derive(Debug, Clone)]
pub struct TypeConstraint {
    info: ConstraintInfo,
    pub ty: ConstructedType,
}

impl TypeConstraint {
    pub fn new(node: NodeRef, ty: ConstructedType) -> Box<dyn Constraint> {
        Box::new(TypeConstraint {
            info: ConstraintInfo::new(node),
            ty,
        })
    }
}

impl Constraint for TypeConstraint {
    fn info(&self) -> &ConstraintInfo {
        &self.info
    }

    fn info_mut(&mut self) -> &mut ConstraintInfo {
        &mut self.info
    }

    fn instantiate(&self, ctx: &mut InstantiateContext<'_>) -> Box<dyn Constraint> {
        let node = ctx.instantiate_node(&self.info.node);
        let ty = ctx.instantiate_type(&Type::Constructed(self.ty.clone()));

        match ty {
            Type::Node(other) => GroupConstraint::new(node, other),
            Type::Constructed(ty) => TypeConstraint::new(node, ty),
        }
    }

    fn run(&mut self, ctx: &mut ConstraintCtx<'_, '_>) -> ConstraintResult {
        ctx.unify(
            Some(Box::new(self.clone())),
            self.info.node.clone(),
            self.ty.clone(),
        );

        ConstraintResult::None
    }
}

File: compiler/src/typecheck/constraints/TypeConstraint.swift
class TypeConstraint: Constraint {
    let type: ConstructedType

    init(_ node: Node, _ type: ConstructedType) {
        self.type = type
        super.init(node)
    }

    override func instantiate(with context: InstantiateContext) -> Constraint {
        let node = context.instantiate(self.node)
        let type = context.instantiate(.constructed(self.type))

        switch type {
        case .node(let other): return GroupConstraint(node, other)
        case .constructed(let type): return TypeConstraint(node, type)
        }
    }

    override func run(with solver: Solver) -> RunResult {
        solver.unify(self.node, self.type, trace: self)
        return .none
    }
}
